src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/SemanticAnalyzer.java

```diff
@@ -0,0 +1,315 @@
+     *      'a.name' -> TEXT
+     *      'a.age' -> INTEGER
+     *      'a.projects' -> NESTED
+     *      'a.projects.name' -> KEYWORD
+     *      'a.projects.active' -> BOOLEAN
+     *      ----- new definitions -----
+     *      'p' -> NESTED
+     *      'p.name' -> KEYWORD
+     *      'p.active' -> BOOLEAN
+     */
+    @Override
+    public Type visitNestedIndexName(String indexName, String alias) {
+        Type type = resolve(new Symbol(Namespace.FIELD_NAME, indexName));
+        if (type != BaseType.NESTED) {
+            throw new SemanticAnalysisException(StringUtils.format(
+                "Field [%s] is [%s] type but nested type is required.", indexName, type));
+        }
+
+        if (!alias.isEmpty()) {
+            redefineFieldNameByPrefixingAlias(indexName, alias);
+        }
+        return NESTED_INDEX;
+    }
+
+    @Override
+    public Type visitIndexPattern(String indexPattern, String alias) {
+        throw new SemanticUnsupportedException("Index pattern is valid but current semantic analyzer cannot parse it");
+    }
+
+    @Override
+    public Type visitFieldName(String fieldName) {
+        // Bypass hidden fields which is not present in mapping, ex. _id, _type.
+        if (fieldName.startsWith("_")) {
+            return UNKNOWN;
+        }
+        return resolve(new Symbol(Namespace.FIELD_NAME, fieldName));
+    }
+
+    @Override
+    public Type visitFunctionName(String funcName) {
+        // Ignore case for function name
+        return resolve(new Symbol(Namespace.FUNCTION_NAME, StringUtils.toUpper(funcName)));
+    }
+
+    @Override
+    public Type visitComparisonOperator(String opName) {
+        return resolve(new Symbol(Namespace.OPERATOR_NAME, StringUtils.toUpper(opName)));
+    }
+
+    @Override
+    public Type visitString(String text) {
+        return BaseType.STRING;
+    }
+
+    @Override
+    public Type visitInteger(String text) {
+        return BaseType.INTEGER;
+    }
+
+    @Override
+    public Type visitFloat(String text) {
+        return BaseType.FLOAT;
+    }
+
+    @Override
+    public Type visitBoolean(String text) {
+        // "IS [NOT] MISSING" can be used on any data type
+        return text.equalsIgnoreCase("MISSING") ? UNKNOWN : BaseType.BOOLEAN;
+    }
+
+    @Override
+    public Type visitDate(String text) {
+        return BaseType.DATE;
+    }
+
+    @Override
+    public Type defaultValue() {
+        return DEFAULT;
+    }
+
+    /** Define by replace with alias */
+    private void redefineFieldNameByPrefixingAlias(String indexName, String alias) {
+        Map<String, Type> typeByFullName = environment().resolveByPrefix(new Symbol(Namespace.FIELD_NAME, indexName));
+        typeByFullName.forEach(
+            (fieldName, fieldType) -> defineFieldName(fieldName.replace(indexName, alias), fieldType)
+        );
+    }
+
+    private void defineFieldName(String fieldName, String type) {
+        defineFieldName(fieldName, BaseType.typeOf(type));
```

Comment:
Just looked if it could be inlined. But it seems more compact with this overloaded method used here and next line followed: https://github.com/dai-chen/sql/blob/improve-semantic-check/src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/SemanticAnalyzer.java#L152